chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc

```diff
@@ -31,25 +33,213 @@ void MaybeClearAccessDeniedException(StorageArea* storage,
+
+  MaybeClearAccessDeniedException(storage, *window, &exception_state);
+  if (base::FeatureList::IsEnabled(features::kBraveEphemeralStorage)) {
+    storage = ephemeralSessionStorage(storage);
+  }
+
+  return storage;
+}
+
+StorageArea* BraveDOMWindowStorage::ephemeralSessionStorage(
+    StorageArea* non_ephemeral_storage) {
+  if (!base::FeatureList::IsEnabled(features::kBraveEphemeralStorage))
+    return non_ephemeral_storage;
+
+  LocalDOMWindow* window = GetSupplementable();
+  if (!window->IsCrossSiteSubframe())
+    return non_ephemeral_storage;
+
+  // If we were not able to create non-ephemeral session storage for this
+  // window, then don't attempt to create an ephemeral version.
+  if (!non_ephemeral_storage)
+    return nullptr;
+
+  if (ephemeral_session_storage_)
+    return ephemeral_session_storage_;
+
+  Page* page = window->GetFrame()->GetDocument()->GetPage();
+  EphemeralStorageNamespaces* namespaces =
+      EphemeralStorageNamespaces::From(page, window);
+  if (!namespaces)
+    return nullptr;
+
+  auto storage_area =
+      namespaces->session_storage()->GetCachedArea(window->GetSecurityOrigin());
+
+  Document* document = window->GetFrame()->GetDocument();
+  ephemeral_session_storage_ =
+      StorageArea::Create(document->GetFrame(), std::move(storage_area),
+                          StorageArea::StorageType::kSessionStorage);
+  return ephemeral_session_storage_;
+}
+
+StorageArea* BraveDOMWindowStorage::localStorage(
+    ExceptionState& exception_state) {
+  LocalDOMWindow* window = GetSupplementable();
+  auto* storage = DOMWindowStorage::From(*window).localStorage(exception_state);
+
+  MaybeClearAccessDeniedException(storage, *window, &exception_state);
+  if (base::FeatureList::IsEnabled(features::kBraveEphemeralStorage)) {
+    storage = ephemeralLocalStorage(storage);
+  }
+
   return storage;
 }
 
+StorageArea* BraveDOMWindowStorage::ephemeralLocalStorage(
+    StorageArea* non_ephemeral_storage) {
+  if (!base::FeatureList::IsEnabled(features::kBraveEphemeralStorage))
+    return non_ephemeral_storage;
+
+  LocalDOMWindow* window = GetSupplementable();
+  if (!window->IsCrossSiteSubframe())
+    return non_ephemeral_storage;
+
+  // If we were not able to create non-ephemeral local storage for this window,
+  // then don't attempt to create an ephemeral version.
+  if (!non_ephemeral_storage)
+    return nullptr;
+
+  if (ephemeral_local_storage_)
+    return ephemeral_local_storage_;
+
+  Page* page = window->GetFrame()->GetDocument()->GetPage();
+  EphemeralStorageNamespaces* namespaces =
+      EphemeralStorageNamespaces::From(page, window);
+  if (!namespaces)
+    return nullptr;
+
+  auto storage_area =
+      namespaces->local_storage()->GetCachedArea(window->GetSecurityOrigin());
+
+  Document* document = window->GetFrame()->GetDocument();
+  ephemeral_local_storage_ =
+      StorageArea::Create(document->GetFrame(), std::move(storage_area),
```

Comment:
just `window->GetFrame()`

Reply:
:+1:

Reply:
Just a note here, since it bit the farbling work recently; will this work in a disconnected frame?  do you need to do the same thing as here? https://github.com/brave/brave-core/pull/6941/files

Reply:
maybe of interest to @iefremov ?

Reply:
Yeah @pes10k I think we are protected from this issue, becasue at first we get "normal" storage from the parent method (e.g. `DOMWindowStorage::sessionStorage()` and it explicitly checks for `!window->GetFrame()`